### PR TITLE
[Clojure] Add Fine Tuning Sentence Pair Classification BERT Example

### PR DESCRIPTION
**Pull request by gigasquid**
## Description ##

This adds a fine tuning example for BERT based off of the following Gluon NLP example

[https://gluon-nlp.mxnet.io/examples/sentence_embedding/bert.html](https://gluon-nlp.mxnet.io/examples/sentence_embedding/bert.html)

There is a PR in the GluonNLP repo for the export script https://github.com/dmlc/gluon-nlp/pull/672

The original BERT infer example directory has been changed from `bert-qa` to `bert` to accomodate both the QA example and this one.

A walkthrough in the form of a Jupyter notebook has been provided for the sentence pair classification based off of the Gluon NLP one. A markdown version of the notebook has also been provided for those who don't have the `lein-jupyter` plugin.

Example results for sentence pair classification after 3 epochs:

```
Speedometer: epoch  2  count  7  metric  [accuracy 0.69921875]
Speedometer: epoch  2  count  8  metric  [accuracy 0.7013889]
Speedometer: epoch  2  count  9  metric  [accuracy 0.690625]
Speedometer: epoch  2  count  10  metric  [accuracy 0.69034094]
Speedometer: epoch  2  count  11  metric  [accuracy 0.6953125]
Speedometer: epoch  2  count  12  metric  [accuracy 0.7019231]
```

which is comparable to the sentence embedding results in the gluon nlp notebook

```
[Epoch 2 Batch 4/18] loss=0.6563, lr=0.0000050, acc=0.638
[Epoch 2 Batch 8/18] loss=0.5937, lr=0.0000050, acc=0.676
[Epoch 2 Batch 12/18] loss=0.5657, lr=0.0000050, acc=0.690
[Epoch 2 Batch 16/18] loss=0.6428, lr=0.0000050, acc=0.674
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Move BERT QA infer example within a broader BERT example
- Add BERT Sentence Pair Classification Example along with notebook
- Tweak the Callback Speedometer so that the results will be printed out in the notebook instead of having to look at the console for the logging.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: contrib/clojure-package/examples/bert/get_bert_data.sh [hunk 1]
```diff
@@ -19,11 +19,14 @@
...
-data_path=model
+data_path=data
 
 if [ ! -d "$data_path" ]; then
   mkdir -p "$data_path"
   curl https://s3.us-east-2.amazonaws.com/mxnet-scala/scala-example-ci/BertQA/vocab.json -o $data_path/vocab.json
   curl https://s3.us-east-2.amazonaws.com/mxnet-scala/scala-example-ci/BertQA/static_bert_qa-0002.params -o $data_path/static_bert_qa-0002.params
   curl https://s3.us-east-2.amazonaws.com/mxnet-scala/scala-example-ci/BertQA/static_bert_qa-symbol.json -o $data_path/static_bert_qa-symbol.json
+  curl https://media.githubusercontent.com/media/gigasquid/mxnet-data/master/static_bert_base_net-symbol.json -o $data_path/static_bert_base_net-symbol.json
+  curl https://media.githubusercontent.com/media/gigasquid/mxnet-data/master/static_bert_base_net-0000.params -o $data_path/static_bert_base_net-0000.params
```
Comment (gigasquid) [inline contrib/clojure-package/examples/bert/get_bert_data.sh]: @lanking520 Would it be possible to host the BERT Base params on on the same bucket as the other `static-bert-qa` params too rather than my github?

Comment (zachgk) [reply contrib/clojure-package/examples/bert/get_bert_data.sh]: @gigasquid I just added them to the bucket for you. You can now find them with the equivalent paths: 
```
https://s3.us-east-2.amazonaws.com/mxnet-scala/scala-example-ci/BertQA/static_bert_base_net-symbol.json
https://s3.us-east-2.amazonaws.com/mxnet-scala/scala-example-ci/BertQA/static_bert_base_net-0000.params
```

Comment (gigasquid) [reply contrib/clojure-package/examples/bert/get_bert_data.sh]: thank you!!

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                   [valid-length]]
+     :label (if (= "0" label)
+              [0]
+              [1])
+     :tokens tokens
+     :train-item train-item}))
+
+(defn fine-tune-model
+  "msymbol: the pretrained network symbol
+    arg-params: the argument parameters of the pretrained model
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: `arg-params` does not seem to be a param to the function here?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+     :train-num (count processed-datas)}))
+
+
+(defn train [dev num-epoch]
+  (let [bert-base (m/load-checkpoint {:prefix model-path-prefix :epoch 0})
+        model-sym (fine-tune-model (m/symbol bert-base) {:num-classes 2 :dropout 0.1})
+        {:keys [data0s data1s data2s labels train-num]} (prepare-data dev)
+        batch-size 32
+        data-desc0 (mx-io/data-desc {:name "data0"
+                                   :shape [train-num seq-length]
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: #nitpick

indentation is off here

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                                   :dtype dtype/FLOAT32
+                                     :layout layout/NT})
+        data-desc1 (mx-io/data-desc {:name "data1"
+                                   :shape [train-num seq-length]
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: same

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                                   :dtype dtype/FLOAT32
+                                     :layout layout/NT})
+        data-desc2 (mx-io/data-desc {:name "data2"
+                                     :shape [train-num]
+                                     :dtype dtype/FLOAT32
+                                     :layout layout/N})
+        label-desc (mx-io/data-desc {:name "softmax_label"
+                                     :shape [train-num]
+                                   :dtype dtype/FLOAT32
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: same

**Comment by gigasquid** [comment]
Thanks for the feedback @Chouffe - I'm working on implementing it today :)

**Comment by gigasquid** [comment]
I think this PR is ready for review. 

Please take a look @Chouffe @kedarbellare @hellonico when you get a chance 👀

File: .gitignore [hunk 6]
```diff
@@ -162,3 +162,9 @@ coverage.xml
 
 # Local CMake build config
 cmake_options.yml
+contrib/clojure-package/examples/bert/.lein-failures
```
Comment (kedarbellare) [inline .gitignore]: are these necessary here or can they be added to the `.gitignore` under `contrib/clojure-package/examples/bert/`?

Comment (gigasquid) [reply .gitignore]: thanks I'll do that :)

**Comment by kedarbellare** [comment]
I haven’t reviewed the PR completely but hopefully I can get to it this weekend

File: contrib/clojure-package/examples/bert/project.clj [hunk 7]
```diff
@@ -16,13 +16,15 @@
 ;;
 
 
-(defproject bert-qa "0.1.0-SNAPSHOT"
-  :description "BERT QA Example"
-  :plugins [[lein-cljfmt "0.5.7"]]
+(defproject bert "0.1.0-SNAPSHOT"
+  :description "BERT Examples"
+  :plugins [[lein-cljfmt "0.5.7"]
+            [lein-jupyter "0.1.16" :exclusions [org.clojure/tools.nrepl org.clojure/clojure org.codehaus.plexus/plexus-utils org.clojure/tools.reader]]]
```
Comment (daveliepmann) [inline contrib/clojure-package/examples/bert/project.clj]: Just FYI I had to comment out the `lein-jupyter` plugin in order to have a working REPL with CIDER version 0.19.0 Raleigh.

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            [org.apache.clojure-mxnet.io :as mx-io]
+            [org.apache.clojure-mxnet.layout :as layout]
+            [org.apache.clojure-mxnet.module :as m]
+            [org.apache.clojure-mxnet.ndarray :as ndarray]
+            [org.apache.clojure-mxnet.optimizer :as optimizer]
+            [org.apache.clojure-mxnet.symbol :as sym]))
+
+(def model-path-prefix "data/static_bert_base_net")
+;; epoch number of the model
+(def epoch 0)
```
Comment (daveliepmann) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Neither `epoch` nor `model-vocab` seem to be used.

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+;; the vocabulary used in the model
+(def model-vocab "data/vocab.json")
+;; the input question
+;; the maximum length of the sequence
+(def seq-length 128)
+
+(defn pre-processing
+  "Preprocesses the sentences in the format that BERT is expecting"
+  [ctx idx->token token->idx train-item]
```
Comment (daveliepmann) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: The `ctx` parameter is never used

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+       (mapv #(nth (:input-batch %) n))
+       (flatten)
+       (into [])))
+
+(defn get-raw-data []
+  (csv/parse-csv (slurp "data/dev.tsv") :delimiter \tab))
+
+(defn prepare-data
+  "This prepares the senetence pairs into NDArrays for use in NDArrayIterator"
+  [dev]
```
Comment (daveliepmann) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: `prepare-data` only uses its parameter `dev` to pass to the `pre-processing` fn, which doesn't use it, so I think `prepare-data` could take 0 arguments.

File: contrib/clojure-package/examples/bert/fine-tune-bert.ipynb [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+   "execution_count": 5,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "He said the foodservice pie business doesn 't fit the company 's long-term growth strategy .\n",
```
Comment (daveliepmann) [inline contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: These two sentences are clearly not semantically equivalent. The double quotes in the source file are messing with the CSV parsing. Investigating now; hope to have a fix shortly.

The *data/dev.tsv* file has it correct:

> Quality	#1 ID	#2 ID	#1 String	#2 String
> 1	1355540	1355592	He said the foodservice pie business doesn 't fit the company 's long-term growth strategy .	" The foodservice pie business does not fit our long-term growth strategy .

Note that strict CSV parsing throws an error:

```clojure
(csv/parse-csv (slurp "data/dev.tsv") :delimiter \tab :strict true)

=>
Unhandled java.lang.Exception
   Double quote present in unquoted field.

                  core.clj:  107  clojure-csv.core/read-unquoted-field
                  core.clj:   93  clojure-csv.core/read-unquoted-field
                  core.clj:  181  clojure-csv.core/parse-csv-line
                  core.clj:  153  clojure-csv.core/parse-csv-line
                  core.clj:  192  clojure-csv.core/parse-csv-with-options/fn
              LazySeq.java:   40  clojure.lang.LazySeq/sval
              LazySeq.java:   49  clojure.lang.LazySeq/seq
                 Cons.java:   39  clojure.lang.Cons/next
                   RT.java:  706  clojure.lang.RT/next
                  core.clj:   64  clojure.core/next
              dispatch.clj:   68  clojure.pprint/pprint-simple-list/fn
              dispatch.clj:   67  clojure.pprint/pprint-simple-list
              dispatch.clj:   77  clojure.pprint/pprint-list
              dispatch.clj:   76  clojure.pprint/pprint-list
              MultiFn.java:  229  clojure.lang.MultiFn/invoke
           pprint_base.clj:  194  clojure.pprint/write-out
           pprint_base.clj:  249  clojure.pprint/pprint/fn
           pprint_base.clj:  248  clojure.pprint/pprint
           pprint_base.clj:  241  clojure.pprint/pprint
           pprint_base.clj:  245  clojure.pprint/pprint
           pprint_base.clj:  241  clojure.pprint/pprint
                  Var.java:  381  clojure.lang.Var/invoke
                pprint.clj:   67  cider.nrepl.middleware.pprint/handle-pprint-fn/fn
                pprint.clj:   85  cider.nrepl.middleware.pprint/pprint-reply/fn
                pprint.clj:   77  cider.nrepl.middleware.pprint/pprint-reply
                pprint.clj:   75  cider.nrepl.middleware.pprint/pprint-reply
                pprint.clj:   95  cider.nrepl.middleware.pprint/pprint-transport/reify
    interruptible_eval.clj:  109  nrepl.middleware.interruptible-eval/evaluate/fn/fn
                  main.clj:  244  clojure.main/repl/read-eval-print
                  main.clj:  261  clojure.main/repl/fn
                  main.clj:  261  clojure.main/repl
                  main.clj:  177  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   83  nrepl.middleware.interruptible-eval/evaluate/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  657  clojure.core/apply
                  core.clj: 1965  clojure.core/with-bindings*
                  core.clj: 1965  clojure.core/with-bindings*
               RestFn.java:  425  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   81  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   50  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  221  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
    interruptible_eval.clj:  189  nrepl.middleware.interruptible-eval/run-next/fn
                  AFn.java:   22  clojure.lang.AFn/run
   ThreadPoolExecutor.java: 1128  java.util.concurrent.ThreadPoolExecutor/runWorker
   ThreadPoolExecutor.java:  628  java.util.concurrent.ThreadPoolExecutor$Worker/run
               Thread.java:  834  java.lang.Thread/run
```

Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: Just removing quotes seems to work on this example. I'm spot-checking the rest.

```clojure
(csv/parse-csv (string/replace (slurp "data/dev.tsv") "\"" "") :delimiter \tab :strict true)
```

Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: I submit a fix in this PR: https://github.com/gigasquid/incubator-mxnet/pull/5

File: contrib/clojure-package/examples/bert/fine-tune-bert.ipynb [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+    "\n",
+    "(def prepared-data {:data0s (slice-inputs-data processed-datas 0)\n",
+    "                    :data1s (slice-inputs-data processed-datas 1)\n",
+    "                    :data2s (slice-inputs-data processed-datas 2)\n",
+    "                    :labels (->> (mapv :label processed-datas)\n",
+    "                                 (flatten)\n",
+    "                                 (into []))\n",
+    "                    :train-num (count processed-datas)})\n",
+    "\n",
+    "(def batch-size 32)\n",
```
Comment (daveliepmann) [inline contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: Defining `batch-size` outside `train-data` is unnecessary, as it is `let`-bound on line 380.

**Comment by gigasquid** [comment]
Thanks for the feedback @daveliepmann - plan to address it shortly :)

File: contrib/clojure-package/examples/bert/project.clj (hunk 7 above)
Comment (gigasquid) [reply contrib/clojure-package/examples/bert/project.clj]: I ran into the some issues as well with cider and the older version of lein-jupyter playing well together. I added a note in the README and in the project.clj

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 8 above)
Comment (gigasquid) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Thanks - I'll add the num-epoch as a command line arg :) and delete the other one

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 10 above)
Comment (gigasquid) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: nice catch

File: contrib/clojure-package/examples/bert/fine-tune-bert.ipynb (hunk 11 above)
Comment (gigasquid) [reply contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: thanks!
